### DISCUSSION TITLE
Transforming structure

### DISCUSSION THREAD

**Discussion by CR-OrenYekutieli** [vectordotdev/vector#21472] [Q&A] [answered]
Hi, I want to covert incoming messages like:
```
{
   "f1": "AABBCC",
   "f2": "DEF",
   "f3": 3,
   "feature": [{
       "name": "nameX",
       "string": "aaaaa"
   }, {
       "name": "nameY",
       "long": 333
   }],
   "f10": 3.14
}
```
to:
```
{
   "f1": "AABBCC",
   "f2": "DEF",
   "f3": 3,
   "fields": {
       "nameX": "aaaaa",
       "nameY": 333
   },
   "f10": 3.14
}
```
that is, only to alter the "feature" field from the above array to an object with the key "fields" that will hold a representation of the array items.
As a newbie to vrl I'm trying to do something like this:
```
result = {}
map_values(.feature) -> |value| {
   key = value.name
   val = "wrong type"
   if exists(value.long) {
     val = value.long
   } else if exists(value.string) {
     val = value.string
   }  # more types here 
   result |= { "{{ key }}" : val }
}
. |= { "fields" : result }
```
The thing is that the "{{ key }}" is not transformed to the variable name and left as "{{ key }}".
What am I missing?

**Comment by jszwedko** [comment, marked as answer]
Hi @CR-OrenYekutieli ! For setting dynamic keys, the `set` function must be used like:

```
result = set!(result,  [key], val }
```

See https://vector.dev/docs/reference/vrl/functions/#set for details.

**Reply by CR-OrenYekutieli** [reply]
Thanks!
